README

PHP-Mesh   - Page meshing framework for PHP.
Copyright (C) 2003-2009  Trejkaz

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

You can contact the author by electronic mail, which is presently
at the following address: trejkaz@trypticon.org


PURPOSE:

    There has been a wonderful project for a while, started by the OpenSymphony
    team, called SiteMesh.

    SiteMesh is a way to merge and style web pages, without needing complicated
    and hard to learn technologies such as XSL, and without needing unnecessary
    cruft such as header and footer includes.

    The goal of this project is to bring the convenience and simplicity of the
    SiteMesh approach to PHP, borrowing from the Java version where appropriate.

REQUIRMENTS:

    PHP 4.1.2 is known to work.  Earlier versions may work too but are untested.
        (I will add more to this list if you send me a report on what works
         and what doesn't.)

    An Apache web server environment is preferable, as it is the server which
    my sample configuration scripts are written for.  Other servers should work
    with the correct configuration (but does anyone really run anything other
    than Apache anymore?)

    For the php_value directives to work in your configuration, the web server
    requires the AllowOverride directive to contain Options.  To be able to
    decorate files whose extension is .html, you will need AllowOverride FileInfo
    as well.

INSTALLING:

    The following instructions will have you install a test page to check whether
    PHP-Mesh will work for you.

    Instructions will assume installation by an ordinary user, not for
    site-wide use.

    1. Extract the PHP-Mesh archive to a known location, such as ~/phpmesh .

    2. Copy the entire contents of the example/ directory to somewhere in your
       web directory.  In my example, I will copy the entire example directory
       into /var/www/localhost/htdocs/example .

    3. Edit the .htaccess file (/var/www/localhost/htdocs/example/.htaccess in
       my case) so that the placeholders point to the correct location where you
       extracted PHP-Mesh.

    4. View the test page in your web browser.  In my case the URL for the test
       page is http://localhost/examples/ .  You should see a red border around
       that page.  If you do not, then you should see an error message detailing
       what might be wrong.

INSTALLATION NOTES:

    On some default Apache installs, including that of Mac OSX, Apache is
    configured to allow no overrides.  In this event, you will need to add
    "AllowOverride Options" to your Apache configuration, so that the php_value
    settings in the .htaccess file work correctly.

FURTHER INFORMATION:

    For further information, I suggest checking out the docs/ directory for more
    detailed documentation.

TODO:

    I need to make some more documented examples, and write a tutorial.

FEEDBACK:

    Feedback and suggestions to trejkaz@trypticon.org




CHANGES

Change Log
-----------

Not yet released:

  - Removed usage of ereg() function, which no longer exists in PHP 6.
  - BUGFIX: Fixed issue parsing the HTTP_ACCEPT_LANGUAGE header which resulted in the
    country part being ignored (so for example, 'en-US' would be stripped to just 'en'.)
  - BUGFIX: Fixed another regex error causing problems for running on Windows systems.

Version 1.4:

  - PHP-Mesh now supports multi-language decorators.  A decorator called "default" will
    result in a lookup for "default.zh-CN.php" if the language is set to zh-CN.  Fallback
    is as always, to "default.php".
  - When running outside of Apache, we now also look at SCRIPT_FILENAME to find out where
    the script is running from, which fixes finding .phpmeshrc on some less friendly web
    hosting setups.
  - BUGFIX: Fixed issue where it would only find the first property.
  - BUGFIX: Fixed formatter output for http-equiv, which would previously print "name="
    instead of "http-equiv=".
  - BUGFIX: Fixed a regex error which was causing problems for running on Windows systems.
  - BUGFIX: Fixed assignment bug which prevented default values working correctly (thanks to
    Dimitri De Franciscis)

Version 1.3:

  - PHP-Mesh no longer requires .phpmeshrc to exist in the document root.  It will search the
    current directory, then the parent, all the way up to the document root.  This is done using
    the Apache functions so it will handle aliased directories as well.

  - The example now works so that the user should only have to copy the example into the web
    server and modify its .htaccess file.

Version 1.2.2:

  - BUGFIX: Fixed incorrect name for $property_name which was making the get_property
    function not work as expected (thanks to a couple of my users for spotting that one.)

Version 1.2.1:

  - BUGFIX: Fix problems with detecting absolute paths on Windows by using the new function
    resolve_path, rather than relying on POSIX conventions.

  - Added utils.php, with some common utilities which I'll need to make things more portable.
    Added the unit tests at the same time.

Version v1.2:

  - FEATURE REQUEST: Query strings are now supported with $page->apply_decorator(), so you can
    now decorate pages like "../portlets/somepage.php?value=5".

  - Added the "identity" decorator to work natively.  decorator=identity in the query string
    now causes all buffering to be bypassed.

Version v1.1:

  - FEATURE REQUEST: Made $REQUEST_URI easily available through $page.

Version v1.0:

  - FEATURE REQUEST: Made regular expressions match the upper-case HTML tags of old.
  
  - FEATURE REQUEST: $DOCUMENT_ROOT variable is not present when server globals are disabled,
    so use $_SERVER[] instead.

Version v0.5:

  - Implemented the 'apply_decorator' function on the Page class, which allows decorators to
    be applied to pages on demand.  Currently implemented only from the decorators themselves,
    but it might be advantageous to allow this on pages even though I can't see a use for it
    just yet.

    This provides portal-like functionality, automatically embedding a page which works
    standalone, such that it will work as a fragment.

  - Abstracted decorator location and execution logic into two new classes: DecoratorSelector
    and Decorator respectively.

  - Changed many method and variable names, bringing them more inline with SiteMesh's tag names
    where appropriate and making naming scheme more PHP-like.


Version 0.1:

  - Initial release.




Page.class.php
<?php
/*
    PHP-Mesh - A page meshing framework for PHP.
    Copyright (C) 2003-2009  Trejkaz

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    You can contact the author by electronic mail, which is presently
    at the following address: trejkaz@trypticon.org
*/

/**
 * This class parses the HTML page which is to be decorated, and provides
 * various convenience methods for getting and printing data from the page.
 *
 * @author Trejkaz <trejkaz@trypticon.org>
 */
class Page
{
    // The page title.
    // This variable is prone to change, so don't use it directly.
    var $_title;

    // The remainder of the page head contents, after the title is removed.
    // This variable is prone to change, so don't use it directly.
    var $_head;

    // The page properties.
    // This variable is prone to change, so don't use it directly.
    var $_properties = array();

    // Maps property name to "name" or "http-equiv" depending on where a meta property came from.
    var $_meta_property_types = array();

    // The page body.
    // This variable is prone to change, so don't use it directly.
    var $_body;

    /**
     * Constructs a Page object.
     *
     * @param $page_content The entire page content, as a single string.
     */
    function Page($page_content)
    {
        // Match the <head/> element.
        if (preg_match("#<head.*?>(.*?)</head>#si", $page_content, $match))
        {
            $this->_head = $match[1];

            // Match the <title/> element.
            if (preg_match("#<title.*?>(.*?)</title>#si", $this->_head, $match))
            {
                // Store away the title.
                $this->_title = trim($match[1]);

                // Match <meta/> tags.
                if (preg_match_all("#<meta.*?>#si", $this->_head, $matches, PREG_SET_ORDER))
                {
                    foreach ($matches as $match)
                    {
                        $meta_tag = $match[0];

                        // Match the key.  (name and http-equiv are treated equally.)
                        if (preg_match("#\b(name)\s*=\s*\"(.*?)\"#si", $meta_tag, $match) ||
                            preg_match("#\b(http-equiv)\s*=\s*\"(.*?)\"#si", $meta_tag, $match))
                        {
                            $meta_type = $match[1];
                            $meta_name = $match[2];
                        }

                        // Match the value.
                        if (preg_match("#\bcontent\s*=\s*\"(.*?)\"#si", $meta_tag, $match))
                        {
                            $meta_value = $match[1];
                        }

                        // Store away the meta key and value.
                        $this->_properties['meta.' . $meta_name] = $meta_value;
                        $this->_meta_property_types['meta.' . $meta_name] = $meta_type;
                    }
                }

                // Store away the header with the title removed.
                $this->_head = preg_replace("#<title.*?>.*?</title>#si", "", $this->_head);
                $this->_head = trim($this->_head);
            }
        }

        // Match the <body/> element.
        if (preg_match("#(<body.*?>)(.*?)</body>#si", $page_content, $match))
        {
            // Store away the body.
            $this->_body = trim($match[2]) . "\n";

            // Match the attributes in the body tag.
            $body_start_tag = $match[1];
            if (preg_match_all("#\b(\S+)\s*=\s*\"(.*?)\"#s", $body_start_tag, $matches, PREG_SET_ORDER))
            {
                foreach ($matches as $match)
                {
                    // Store away the body attribute key and value.
                    $this->_properties['body.' . $match[1]] = $match[2];
                }
            }
        }
    }

    /**
     * Print the page title.
     *
     * @param $default The default value of the title, to use if no property is defined.
     */
    function title()
    {
        print($this->_title);
    }

    /**
     * Print the remainder of the head contents, after the title is removed.
     */
    function head()
    {
        print($this->_head);
    }

    /**
     * Gets the URL of the page which is being viewed.
     *
     * @return the URL of the page which is being viewed.
     */
    function get_url()
    {
        return $_SERVER['REQUEST_URI'];
    }

    /**
     * Get a single property.
     *
     * @param $property_name The name of the property.
     * @param $default The default value of the property, to use if the property is not defined.
     * @return the value of that property, or the default if the property was not defined.
     */
    function get_property($property_name, $default = NULL)
    {
        $property_value = $this->_properties[$property_name];
        if ($property_value == NULL)
        {
            $property_value = $default;
        }
        return $property_value;
    }

    /**
     * Print a single property.
     *
     * Properties found in <meta> tags are named meta.[propertyName].
     * Properties found on the <body> tag as attributes are named body.[propertyName].
     *
     * @param $property_name The name of the property.
     * @param $default The default value of the property, to use if the property is not defined.
     * @param $formatted Whether to format the property as it originally appeared.
     */
    function property($property_name, $default = NULL, $formatted = false)
    {
        $property_value = $this->get_property($property_name, $default);
        if ($property_value != NULL)
        {
            if ($formatted)
            {
                if (preg_match("/^(body|meta).(.*)$/", $property_name, $match))
                {
                    $property_prefix = $match[1];
                    $property_name_cut = $match[2];
                }
                else
                {
                    die("Invalid property name: $property_name.\n");
                }

                if ($property_prefix == "meta")
                {
                    $meta_type = $this->_meta_property_types[$property_name];
                    $before = "<meta $meta_type=\"$property_name_cut\" content=\"";
                    $after = "\" />";
                }
                else if ($property_prefix == "body")
                {
                    $before = " $property_name_cut=\"";
                    $after = "\"";
                }
                else
                {
                    die("Impossible branch: $property_name   =>   $property_prefix $property_name_cut");
                }
            }
            else
            {
                $before = "";
                $after = "";
            }

            print "$before$property_value$after";
        }
    }

    /**
     * Print the page body.
     */
    function body()
    {
        print($this->_body);
    }

    /**
     * Includes the specified page as decorated content.
     *
     * @param $page_location the location of the page to fetch and apply the
      *                      decorator to, which may be relative to the decorator.
     * @param $decorator_name the name of the decorator to decorate the page with.
     */
    function apply_decorator($page_location, $decorator_name)
    {
        global $decorator_selector;
        global $_GET;
        global $_POST;

        // Store the original GET and POST variables.
        $OLD_GET = $_GET;
        $OLD_POST = $_POST;

        // Does the page location have a query string?  If it does, we need to parse it!
        $_GET = array();
        $_POST = array();
        $parsed_location = parse_url($page_location);
        if (isset($parsed_location['query']))
        {
            parse_str($parsed_location['query'], $_GET);
        }

        // Capture the output from requiring the page.
        ob_start();
        require($decorator_selector->get_path($parsed_location['path']));
        $page_contents = ob_get_clean();

        // Restore the original GET and POST variables.
        $_GET = $OLD_GET;
        $_POST = $OLD_POST;

        // Create the page object.  This guy does all the parsing work.
        $page = new Page($page_contents);

        // Create the decorator.
        $decorator = $decorator_selector->get_decorator($decorator_name);

        // Perform the decoration.
        $decorator->decorate($page);
    }

} // class Page

?>
